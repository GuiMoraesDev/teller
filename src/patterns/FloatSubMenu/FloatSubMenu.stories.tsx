import React from 'react';

import { ComponentMeta, ComponentStory } from '@storybook/react';

import FloatSubMenu from '.';

const Storie = {
	title: 'Patterns/FloatSubMenu',
	component: FloatSubMenu,
} as ComponentMeta<typeof FloatSubMenu>;

export default Storie;

const Template: ComponentStory<typeof FloatSubMenu> = (args) => (
	<FloatSubMenu {...args} />
);

export const Default = Template.bind({});

Default.args = {
	ActionComponent: <span>Controller</span>,
	submenuComponents: [
		<p key="menu-item-1">Menu item 1</p>,
		<p key="menu-item-2">Menu item 2</p>,
		<p key="menu-item-3">Menu item 3</p>,
	],
};
