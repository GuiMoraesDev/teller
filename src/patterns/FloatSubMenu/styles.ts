import styled, { css } from 'styled-components';

import { StyledProps } from '.';

export const Container = styled.div<StyledProps>`
	position: relative;

	display: flex;

	place-content: center;

	height: 100%;

	${({ isFullWidth }) =>
		isFullWidth &&
		css`
			width: 100%;
		`};
`;

export const ActionButton = styled.button`
	position: relative;

	height: 100%;

	border: none;
`;

interface NavWrapperProps {
	parentWidth?: number;
	isVisible: boolean;
}

export const NavWrapper = styled.nav<NavWrapperProps>`
	position: absolute;

	bottom: 0;
	left: 50%;

	background-color: ${({ theme }) => theme.themeColors.canvasDark};

	border: ${({ theme }) => theme.borders.solid};
	border-radius: ${({ theme }) => theme.rounded.sm};

	box-shadow: ${({ theme }) => theme.shadows.regular};

	transform: translate(
		-20%,
		calc(100% + ${({ theme }) => theme.sizes.common.x2})
	);

	visibility: ${({ isVisible }) => (isVisible ? 'visible' : 'hidden')};

	:before {
		content: '';

		position: absolute;

		top: 0;
		left: calc(${({ parentWidth }) => parentWidth}px / 3);

		width: 0;
		height: 0;

		border: ${({ theme }) => theme.borders.solid};
		border-width: ${({ theme }) => theme.sizes.common.x2};
		border-color: transparent;
		border-bottom-color: ${({ theme }) => theme.themeColors.canvasDark};

		transform: translateY(-100%);
	}
`;
