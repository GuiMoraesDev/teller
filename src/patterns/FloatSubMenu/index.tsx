import { useCallback, useRef, useState, useEffect } from 'react';

import * as Styles from './styles';

export interface StyledProps {
	isFullWidth?: boolean;
}

interface Props extends StyledProps {
	ActionComponent: JSX.Element;
	submenuComponents: JSX.Element[];
}

const FloatSubMenu = ({
	ActionComponent,
	submenuComponents,
	isFullWidth = false,
}: Props): JSX.Element => {
	const containerRef = useRef<HTMLDivElement>(null);
	const actionButtonRef = useRef<HTMLButtonElement>(null);
	const navContainerRef = useRef<HTMLDivElement>(null);

	const [visible, setVisible] = useState(false);

	const handleToggleSubmenu = useCallback(() => {
		setVisible((state) => !state);
	}, []);

	const handleClickOutside = useCallback((event: MouseEvent) => {
		const clickedOnActionButton = actionButtonRef.current?.contains(
			event.target as Node
		);
		const clickedOnNavContainer = navContainerRef.current?.contains(
			event.target as Node
		);

		if (!clickedOnActionButton && !clickedOnNavContainer) {
			setVisible(false);
		}
	}, []);

	useEffect(() => {
		if (visible) {
			document.addEventListener('mousedown', handleClickOutside);

			return () => {
				document.removeEventListener('mousedown', handleClickOutside);

				setVisible(false);
			};
		}
	}, [visible, handleClickOutside]);

	return (
		<Styles.Container
			data-testid="float-container"
			ref={containerRef}
			isFullWidth={isFullWidth}
		>
			<Styles.ActionButton
				onClick={handleToggleSubmenu}
				ref={actionButtonRef}
				data-testid="action-button"
			>
				{ActionComponent}
			</Styles.ActionButton>

			<Styles.NavWrapper
				isVisible={visible}
				ref={navContainerRef}
				data-testid="nav-wrapper"
				parentWidth={actionButtonRef.current?.offsetWidth}
			>
				{submenuComponents.map((Component) => {
					return Component;
				})}
			</Styles.NavWrapper>
		</Styles.Container>
	);
};

export default FloatSubMenu;
