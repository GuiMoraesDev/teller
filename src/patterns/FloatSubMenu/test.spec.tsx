import React from 'react';

import userEvent from '@testing-library/user-event';

import { render, waitFor } from 'tests/test.utils';

import FloatSubMenu from './index';

const ContainerRender = () =>
	render(
		<FloatSubMenu
			ActionComponent={<span>Controller</span>}
			submenuComponents={[
				<p key="menu-item-1">Menu item 1</p>,
				<p key="menu-item-2">Menu item 2</p>,
				<p key="menu-item-3">Menu item 3</p>,
			]}
		/>
	);

const elementsTestId = {
	floatContainer: 'float-container',
	actionButton: "action-button",
	navWrapper: "nav-wrapper"
};

describe('FloatSubMenu pattern', () => {
	it('should render correctly', () => {
		const { container, getByTestId } = ContainerRender();

		const floatContainer = getByTestId(elementsTestId.floatContainer);
		const actionButton = getByTestId(elementsTestId.actionButton);
		const navWrapper = getByTestId(elementsTestId.navWrapper);

		expect(floatContainer).toBeVisible();
		expect(actionButton).toBeVisible();
		expect(navWrapper).not.toBeVisible();

		expect(container.firstChild).toMatchSnapshot();
	});

	it('should open menu on action button click', async () => {
		const { getByTestId } = ContainerRender();

		const actionButton = getByTestId(elementsTestId.actionButton);
		const navWrapper = getByTestId(elementsTestId.navWrapper);

		await waitFor(() => userEvent.click(actionButton));

		expect(navWrapper).toBeVisible();
	});

	it('should close menu clicking on another element then action button or nav wrapper', async () => {
		const { getByTestId } = ContainerRender();

		const floatContainer = getByTestId(elementsTestId.floatContainer);
		const actionButton = getByTestId(elementsTestId.actionButton);
		const navWrapper = getByTestId(elementsTestId.navWrapper);

		await waitFor(() => userEvent.click(actionButton));

		expect(navWrapper).toBeVisible();

		await waitFor(() => userEvent.click(floatContainer));

		expect(navWrapper).not.toBeVisible();
	});

	it('should open and close menu clicking on action button', async () => {
		const { getByTestId } = ContainerRender();

		const floatContainer = getByTestId(elementsTestId.floatContainer);
		const actionButton = getByTestId(elementsTestId.actionButton);
		const navWrapper = getByTestId(elementsTestId.navWrapper);

		await waitFor(() => userEvent.click(actionButton));

		expect(navWrapper).toBeVisible();

		await waitFor(() => userEvent.click(actionButton));

		expect(navWrapper).not.toBeVisible();
	});
});
