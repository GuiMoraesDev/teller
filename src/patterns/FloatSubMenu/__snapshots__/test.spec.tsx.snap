// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FloatSubMenu pattern should render correctly 1`] = `
.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  place-content: center;
  height: 100%;
}

.c1 {
  position: relative;
  height: 100%;
  border: none;
}

.c2 {
  position: absolute;
  bottom: 0;
  left: 50%;
  background-color: #E4E5E3;
  border: 0.1rem solid #C9CAC8;
  border-radius: 0.4rem;
  box-shadow: 0px 3px 6px 1px #E4E5E3;
  -webkit-transform: translate( -20%, calc(100% + 0.8rem) );
  -ms-transform: translate( -20%, calc(100% + 0.8rem) );
  transform: translate( -20%, calc(100% + 0.8rem) );
  visibility: hidden;
}

.c2:before {
  content: '';
  position: absolute;
  top: 0;
  left: calc(px / 3);
  width: 0;
  height: 0;
  border: 0.1rem solid #C9CAC8;
  border-width: 0.8rem;
  border-color: transparent;
  border-bottom-color: #E4E5E3;
  -webkit-transform: translateY(-100%);
  -ms-transform: translateY(-100%);
  transform: translateY(-100%);
}

<div
  class="c0"
  data-testid="float-container"
>
  <button
    class="c1"
    data-testid="action-button"
  >
    <span>
      Controller
    </span>
  </button>
  <nav
    class="c2"
    data-testid="nav-wrapper"
  >
    <p>
      Menu item 1
    </p>
    <p>
      Menu item 2
    </p>
    <p>
      Menu item 3
    </p>
  </nav>
</div>
`;
